torch/_inductor/ir.py

```diff
@@ -4585,6 +4585,86 @@ def create(
         )
 
 
+class QMaxpool2dPT2E(ExternKernelAlloc):
+    def __init__(
+        self,
+        layout,
+        inputs,
+        constant_args=(),
+    ):
+        super().__init__(layout, inputs, constant_args)
+
```

Comment:
Why do we need this registration ? Could we instead add `requires_channels_last` here https://github.com/pytorch/pytorch/blob/5e7e8f3cb26b264cf4c16cd8b800fd1f3df12049/torch/_inductor/lowering.py#L1776 and register it that way?

Reply:
Hi @eellison, thanks for the suggestions. Did you mean register it something like `make_fallback(quantized.max_pool2d, requires_channels_last)`

Reply:
Yes, that is what I mean. Is that any functionally different then what is occurring now ?

Reply:
Thanks for suggestion. I think it works and can save lots of duplicated code. Please kindly help to take a look again.

Reply:
Hi @eellison, after I change the implementation to `make_fallback(quantized.max_pool2d, requires_channels_last)`, the test case in this PR passed. However, I have met error in another test case here: https://gist.github.com/leslie-fang-intel/53eadd45b8b451eac2ebc5c37051614c (which works fine if I keep the original implementation).

- The error message is here: https://gist.github.com/leslie-fang-intel/bb6dcb221dabdac4971bc09054da64c9
- The error can be produced by this PyTorch Branch: https://github.com/leslie-fang-intel/pytorch/tree/leslie/enable_cat_recipe

I have done some triage with PDB, and the call stack is confused:

- The last python frame is https://github.com/pytorch/pytorch/blob/9e301949ecdae25f16fc115261141546e16fda2e/torch/_subclasses/fake_tensor.py#L1694 with `func` of `<OpOverload(op='quantized.max_pool2d', overload='default')>`
- Then, with step by step execution, I find it enter into  https://github.com/pytorch/pytorch/blob/9e301949ecdae25f16fc115261141546e16fda2e/torch/_ops.py#L732 and the `op_name` changes to `q_zero_point`. I don't think `aten.q_zero_point.default` should exist in the graph.

Appreciate if you have any clues for next step triage.

Reply:
Oh, I think I may know why:

- Here: https://github.com/pytorch/pytorch/blob/d4d086ce7b53c92db6963caac22eb8761b4b0d44/torch/_subclasses/fake_tensor.py#L1691 creates the `real_tensor` with `contiguous` memory format.
- However, `quantized::max_pool2d` with uint8 data type only support `channels_last` input. It will causes error in the real kernel execution.

Reply:
Hi @eellison, after some study, I think `make_fallback(quantized.max_pool2d, requires_channels_last)` doesn't work as expected to convert inputs to `channels_last` format.

- `make_fallback(quantized.max_pool2d, requires_channels_last)` will add op `quantized.max_pool2d` into `layout_constraints` dict.
- However, when `layout_constraints` dict used here https://github.com/pytorch/pytorch/blob/c287262b02b0284326ed1816e057ac7d80baf38b/torch/_inductor/graph.py#L703-L705. We will get `n.target` as `functools.partial(<function _register_quantized_maxpool2d_lowering.<locals>.qmaxpool2d at 0x7f22dd253b80>, ...)` instead of `quantized.max_pool2d`. Because we only registered the pattern `dq->max_pool2d_with_indices->q` through `register_lowering_pattern`. So, it will not trigger the layout convert function.

I have draft a fix here https://github.com/pytorch/pytorch/pull/106689 to manfully invoke the `require_channels_last` in the lowering of  `dq->max_pool2d_with_indices->q` pattern and before the invoking of `run_fallback_kernel`. Does it look reasonable to you? If so, I can merge it into this PR.